network/tools/network-policy-enforcement-latency/pkg/client.go

```diff
@@ -0,0 +1,552 @@
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			handlePodEvent(obj, true)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			handlePodEvent(newObj, false)
+		},
+	})
+
+	c.podInformer = informer
+	go informer.Run(c.informerStopChan)
+
+	err := utils.Retry(10, 500*time.Millisecond, func() error {
+		return utils.InformerSynced(informer.HasSynced, "pod informer")
+	})
+
+	return err
+}
+
+// processPodCreationEvents is run by each worker to process the incoming pod
+// watch events.
+func (c *TestClient) processPodCreationEvents() {
+	wait.Until(c.processNextPodCreationEvent, 0, c.informerStopChan)
+	klog.Infof("processPodCreationEvents() goroutine quitting")
+}
+
+// processNextPodCreationEvent processes the next pod watch event in the work
+// queue, by sending traffic to the pods that have IP address assigned, until
+// the first successful response.
+func (c *TestClient) processNextPodCreationEvent() {
+	item, quit := c.podCreationWorkQueue.Get()
+	if quit {
+		close(c.informerStopChan)
+		c.podCreationWorkQueue.ShutDown()
+		return
+	}
+	defer c.podCreationWorkQueue.Done(item)
+
+	startTime := time.Now()
+	podEvent, ok := item.(podEvent)
+	if !ok {
+		klog.Warningf("Failed to convert work queue item of type %T to string key", item)
+		return
+	}
+
+	// Skip Update pod events for pods that aren't tracked since their creation.
+	// Only new pods should be included as targets.
+	if c.skipUpdateForUntrackedPods(podEvent) {
+		return
+	}
+
+	// Get pod from informer's cache.
+	objList, err := c.podInformer.GetIndexer().ByIndex(utils.NameIndex, podEvent.podName)
+	if err != nil {
+		klog.Warningf("Failed to get pod object from the informer's cache, for pod name %q", podEvent.podName)
+		return
+	}
+
+	if len(objList) < 1 {
+		klog.Warningf("Pod object does not exist in the informer's cache, for pod name %q", podEvent.podName)
+		return
+	}
+
+	pod, ok := objList[0].(*corev1.Pod)
+	if !ok {
+		klog.Warningf("processNextPodCreationEvent() failed to convert obj (%T) to *corev1.Pod", objList[0])
+		return
+	}
+
+	if len(pod.Status.PodIP) > 0 {
+		podName := pod.GetName()
+		podCreationTime := pod.GetCreationTimestamp().Time
+		haveMaxTargets := false
+
+		c.targetPodsMap.lock.Lock()
+		_, targetUsed := c.targetPodsMap.mp[podName]
+		if !targetUsed {
+			c.targetPodsMap.mp[podName] = startTime
+			haveMaxTargets = len(c.targetPodsMap.mp) > c.config.maxTargets
+		}
+		c.targetPodsMap.lock.Unlock()
+
+		// Process only up to the expected number of pods.
+		if haveMaxTargets {
+			c.stopInformerChan()
+			return
+		}
+
+		// Do the measurements only if it hasn't already been done for this pod.
+		if !targetUsed {
+			target := &targetSpec{
+				ip:        pod.Status.PodIP,
+				name:      podName,
+				namespace: pod.GetNamespace(),
+				startTime: startTime,
+			}
+			go c.recordFirstSuccessfulRequest(target, true)
```

Comment:
You don't want to wait for it?

Reply:
No. It contains the functionality of recording the first successful request, by generating metrics.